### PR TITLE
War Ops, reinforcements, and Syndicate Uplink tweaks. Bounty #6148

### LINKED ISSUE #5556: Basic telecrystal items.
### Description
Telecrystals aren't just the currency used to buy items from the uplink in /tg/, they're also physical items that can be used. In their basic form, telecrystal items can be bought from the uplinks. Each physical telecrystal costs one TC, of course. These physical telecrystals can be used on any item that has an uplink to add TC to said uplink.

Physical telecrystals are useful in /tg/ as a form of transferring TC between uplinks belonging to different traitors and nuclear operatives. Telecrystals also affect the properties of certain traitor items, but that's for another time.

### LINKED ISSUE #6148: War Ops, reinforcements, and Syndicate Uplink tweaks bounty B:90
**This is a bounty! This issues calls for a couple additional features to the Nuke Ops game mode. First It should introduce the ability for Nuclear Operative to declare war on the station, which gives a large TC bonus, but makes the presence of the Nuclear Operatives known to the station and forces them to wait for a period of time, which means that the station has time to mount a defense.

This issue also calls for some tweaks and design additions to the Syndicate Uplink we have in game and also to add syndicate operative reinforcements as a purchaseable option in the TC uplink**

### Design Considerations:

when nukies declare war, a message is broadcast to the station that war has been declared. From there, nukies received a trickle of extra TC on a dedicated telecrystal upload/receive station. These Telecrystals (TC) can be distributed across the nukie team to allow them to buy signifcantly more items than they normally can. 

The bonus TC takes a while to be deposited on the on the upload/receive console, giving the station time to prepare for the nukie attack. _Given how Unitystation handles shuttles and shuttle travelling differently than ss13, this is how the extra TC should be gotten from declaring war. The immobility of the upload/receive station forces the nukies to wait at the syndicate base to get their rewards for declaring war_

This issue requires a few changes to the current implementation of the syndicate TC uplink. First, TC must be shareable, so Nukies can divide up, store, trade, and spend TC flexibly. 

Second, Nukies and Traitors need to have their TC stores separated so that Nukies can have access to the much more powerful and devastating equipment that is not appropriate for traitors.

In addition, this issue should also see the addition of the nukie reinforcement teleporter. The reinforcement teleporter is a purchaseable uplink item that when deployed, allows a single ghost to become a new nukie using the ghost role system. _This nukie comes with standard equipment, but no TC of their own and must be supplied it by other generous Nukies._

Lastly, it needs to be decided whether the Nukie Officer should be brought back as a feature or not. The method of declaring war could be handled by the Officer, but it could also just as easily be integrated into the syndicate console or the upload receive station.

### What is expected to fulfill the bounty:

in order to fulfill the bounty this is what is required:

- [x] A method for Nukies to declare war, either on the syndicate operations console or on the upload/receive station
- [x] TC upload/Receive station that allows nukies to deposit and withdraw TC like a money machine. 
- [x] Physical, shareable Telecrystals
- [x] Separated nukie and traitor TC uplinks
- [x] Nukie reinforcement - both the ghost role and their teleporters
- [x] weapon bundles should be added to the nukie TC uplink, as well as any other nukie items that can be added (this should not require any new code, and could be handled by maintainers)
- [x] all new mapping features need to be mapped into the appropriate scenes (should be easy)

### PR DESCRIPTION
**Pull request by AutumnNova**
- [x]  A method for Nukies to declare war, either on the syndicate operations console or on the upload/receive station
 
 - [x]  TC upload/Receive station that allows nukies to deposit and withdraw TC like a money machine.

 
 - [x]  Physical, shareable Telecrystals
 
 - [x]  Separated nukie and traitor TC uplinks
 
 - [x]  Nukie reinforcement - both the ghost role and their teleporters
 
 - [x]  weapon bundles should be added to the nukie TC uplink, as well as any other nukie items that can be added (this should not require any new code, and could be handled by maintainers)
 
 - [x]  all new mapping features need to be mapped into the appropriate scenes (should be easy)

CL: Implement Warops, reinforcement teleporter and nuke op exclusive uplink items

closes #5556 
closes #6148

### CODE REVIEW

**Comment by corp-0** [comment]
can you solve the conflicts so I review? It should be easy enough to be honest, those files were just moved out of Resources.

**Comment by AutumnNova** [comment]
> can you solve the conflicts so I review? It should be easy enough to be honest, those files were just moved out of Resources.

done

File: UnityProject/Assets/Scripts/Items/PDA/PDALogic.cs [hunk 1]
```diff
@@ -56,6 +56,12 @@ public class PDALogic : NetworkBehaviour, ICheckedInteractable<HandApply>, IChec
 		[SerializeField, BoxGroup("Uplink"), Range(0, 60)]
 		private float informUplinkCodeDelay = 10;
 
+		private bool isNukeOps = false;
+		public bool IsNukeOps => isNukeOps;
+
+		[SerializeField]
+		private ItemTrait trait;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Items/PDA/PDALogic.cs]: this field is not set in any of the PDA's I could find, is that intended? From what I can read it is used to determine if something can be inserted in the PDA.

File: UnityProject/Assets/Scripts/Items/PDA/PDALogic.cs [hunk 2]
```diff
@@ -338,11 +364,12 @@ private void ServerInsertItem(GameObject item, ItemSlot fromSlot)
 		/// </summary>
 		/// <param name="informPlayer">The player that will be informed the code to the PDA uplink</param>
 		/// <param name="tcCount">The amount of telecrystals to add to the uplink.</param>
-		public void InstallUplink(ConnectedPlayer informPlayer, int tcCount)
+		public void InstallUplink(ConnectedPlayer informPlayer, int tcCount, bool isNukie)
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Items/PDA/PDALogic.cs:368]: add the new parameter to the docstring above.

File: UnityProject/Assets/Scripts/Items/PDA/PDALogic.cs (hunk 1 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Items/PDA/PDALogic.cs]: consider naming it something easier to understand.

File: UnityProject/Assets/Scripts/Managers/CentComm.cs [hunk 3]
```diff
@@ -172,7 +172,7 @@ public void ChangeAlertLevel(AlertLevel toLevel, bool announce = true)
 		{
 			if (CurrentAlertLevel == toLevel) return;
 
-			if (CurrentAlertLevel > toLevel && toLevel == AlertLevel.Green)
+			if (CurrentAlertLevel > toLevel && toLevel == AlertLevel.Green && announce)
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Managers/CentComm.cs:175]: is this to stop new rounds from announcing the green level before they start?

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	public int TcIncrement = 14;
+
+	private bool warDeclared = false;
+	private bool rewardGiven = false;
+	private int timer = 1200;
+
+	private int timerIncrement = 60;
+	private int tcToGive = 280;
+
+	public List<SpawnedAntag> Operatives = new List<SpawnedAntag>();
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: looks like this is not meant to be set in editor nor kept after rounds. Add the ``[NonSerialized]`` decorator.

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+using System;
+using System.Collections.Generic;
+
+public class SyndicateOpConsole : MonoBehaviour
+{
+
+
+	public static SyndicateOpConsole Instance;
+
+	public int TcReserve;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: looks like this is not meant to be set in editor nor kept after rounds. Add the [NonSerialized] decorator.

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+	public int TcIncrement = 14;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: I don't understand what the increment is.

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	private void OnDisable()
+	{
+		if (CustomNetworkManager.IsServer)
+		{
+			UpdateManager.Remove(CallbackType.PERIODIC_UPDATE, ServerUpdateTimer);
+		}
+	}
+
+
+	public void AnnounceWar(string DeclerationMessage)
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]:
```suggestion
	public void AnnounceWar(string declarationMessage)
```

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	{
+		if (warDeclared == false)
+		{
+			warDeclared = true;
+
+			GameManager.Instance.CentComm.ChangeAlertLevel(CentComm.AlertLevel.Red, true);
+			CentComm.MakeAnnouncement(ChatTemplates.PriorityAnnouncement, 
+			$"Attention all crew! An open message from the syndicate has been picked up on local radiowaves! Message Reads:\n" +
+			$"{DeclerationMessage}" ,CentComm.UpdateSound.Alert);
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]:
```suggestion
			$"{declarationMessage}" ,CentComm.UpdateSound.Alert);
```

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+			var antagplayers = AntagManager.Instance.CurrentAntags;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]:
```suggestion
			var antagPlayers = AntagManager.Instance.CurrentAntags;
```

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+			foreach (var antag in antagplayers )
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]:
```suggestion
			foreach (var antag in antagPlayers )
```

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+			{
+				if (antag.Antagonist.AntagJobType == JobType.SYNDICATE)
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: was the problem with JobType fixed?

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+	public int TcReserve;
+
+	public int TcIncrement = 14;
+
+	private bool warDeclared = false;
+	private bool rewardGiven = false;
+	private int timer = 1200;
+
+	private int timerIncrement = 60;
+	private int tcToGive = 280;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: make this a serialized field so we can tweak it from editor.

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+				}
+			}
+		}
+	}
+
+	public void ServerUpdateTimer()
+	{
+		if (warDeclared == false || rewardGiven) return;
+
+		if (timerIncrement > 0 || timer > 0)
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: this timer is weird, why the need for two variables? Can't we just count seconds and update the UI or whatever this is used with division and modulo?

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateTelecrystalStation.cs [hunk 14]
```diff
@@ -0,0 +1,38 @@
+using UnityEngine;
+using Items.PDA;
+
+class SyndicateTelecrystalStation : MonoBehaviour,  ICheckedInteractable<HandApply>, IExaminable
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateTelecrystalStation.cs]: add a namespace

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs [hunk 15]
```diff
@@ -0,0 +1,112 @@
+using UnityEngine;
+using Managers;
+using Antagonists;
+using Strings;
+using System;
+using System.Collections.Generic;
+
+public class SyndicateOpConsole : MonoBehaviour
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: add a namespace

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateTelecrystalStation.cs [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+			int tc = Mathf.FloorToInt(SyndicateOpConsole.Instance.Operatives.Count / SyndicateOpConsole.Instance.TcIncrement);
+			pdaComp.UplinkTC += tc;
+			SyndicateOpConsole.Instance.TcReserve -= tc;
+		}
+	}
+	
+	public string Examine(Vector3 vector)
+	{
+		var examine = $"It seems to have {SyndicateOpConsole.Instance.TcReserve} telecrystals in reserve";
+		return examine;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateTelecrystalStation.cs]: return the string directly, no need for the assignation.

File: UnityProject/Assets/Scripts/Objects/ReinforcementTeleporter.cs [hunk 17]
```diff
@@ -0,0 +1,71 @@
+using UnityEngine;
+using Mirror;
+using Systems.GhostRoles;
+using System;
+using System.Collections;
+using ScriptableObjects;
+
+public class ReinforcementTeleporter : MonoBehaviour, ICheckedInteractable<HandActivate>
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/ReinforcementTeleporter.cs:8]: add namespace

File: UnityProject/Assets/Scripts/Systems/Antagonists/AntagManager.cs [hunk 18]
```diff
@@ -27,6 +27,8 @@ public class AntagManager : MonoBehaviour
 		/// </summary>
 		private List<SpawnedAntag> ActiveAntags = new List<SpawnedAntag>();
 
+		public List<SpawnedAntag> CurrentAntags => ActiveAntags;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Systems/Antagonists/AntagManager.cs]: rename the private to use camelCase and make yours be PascalCase

File: UnityProject/Assets/Scripts/UI/Items/PDA/GUI_PDAUplinkItem.cs [hunk 19]
```diff
@@ -24,10 +24,16 @@ public void OnPageDeactivated()
...
+			bool isNukie = controller.mainController.PDA.IsNukeOps;
+			int skipped = 0;
 			for (int i = 0; i < category.ItemList.Count; i++)
 			{
-				dynamicList.Entries[i].GetComponent<GUI_PDAUplinkItemTemplate>().ReInit(category.ItemList[i]);
+				if (isNukie || category.ItemList[i].IsNukeOps == false)
+				{
+					dynamicList.AddItem();
+					dynamicList.Entries[i].GetComponent<GUI_PDAUplinkItemTemplate>().ReInit(category.ItemList[i]);
+					skipped++;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/UI/Items/PDA/GUI_PDAUplinkItem.cs]: I don't think you're using this value anywhere

**Review by corp-0 — COMMENTED** [review]
Just minor stuff, looks good.

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs (hunk 13 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: What I really meant is, this should make its intention clearer.

File: UnityProject/Assets/Scripts/Items/PDA/PDALogic.cs (hunk 1 above)
Comment (AutumnNova) [reply UnityProject/Assets/Scripts/Items/PDA/PDALogic.cs]: this is used to get reference to the telecrystal trait
figured it was better then throwing it in common traits as its only really used here

File: UnityProject/Assets/Scripts/Managers/CentComm.cs (hunk 3 above)
Comment (AutumnNova) [reply UnityProject/Assets/Scripts/Managers/CentComm.cs:175]: yep, im unsure as to why that was changed but removing it just makes it ignore the 'announce' bool

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs (hunk 6 above)
Comment (AutumnNova) [reply UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: 280/20 = 14
its the amount of tc to award every time increment, which in this case is 1 minute

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs (hunk 13 above)
Comment (AutumnNova) [reply UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: the timer variable is used directly for the gui
the timer increment var is used to count to 60

File: UnityProject/Assets/Scripts/UI/Items/PDA/GUI_PDAUplinkItem.cs (hunk 19 above)
Comment (AutumnNova) [reply UnityProject/Assets/Scripts/UI/Items/PDA/GUI_PDAUplinkItem.cs]: ah bugger, your right
meant to remove that but forgot

File: UnityProject/Assets/Scripts/Items/PDA/PDALogic.cs (hunk 1 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Items/PDA/PDALogic.cs]: wonderful, just rename it to something clearer and remember to set the value in the PDA prefabs

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs (hunk 6 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: I see

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs (hunk 13 above)
Comment (corp-0) [reply UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: what happens when you finish counting to 60?

Comment (PerfectTangent) [reply UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: I assume tc gets paid out

Comment (AutumnNova) [reply UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: it pays out TC equivelent to Tc Increment
unless there is not enough tc to give then itll just give the rest of the tc in reserve

Comment (corp-0) [reply UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: Oh yeah, I can see that in the else. In plain English, how should the timer work?

Comment (AutumnNova) [reply UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: one timer counts down from 1200 and the other counts down from 60
probably should use modulo or smth instead

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs (hunk 11 above)
Comment (AutumnNova) [reply UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: yea, nuke ops now have their jobtype assigned in their spawn script

**Comment by AutumnNova** [comment]
> There's a sniper bundle item prefab in the populators. Should that be there?

accidentally commited that instead of the populator, fixed now tho

File: UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs [hunk 20]
```diff
@@ -5,108 +5,105 @@
...
+			{
+				RewardTelecrystals();
+			}
 		}
-		if (tcToGive < TcIncrement)
+		public void RewardTelecrystals()
 		{
-			TcReserve += tcToGive;
-			tcToGive = 0;
+			if (tcToGive <= TcIncrement) rewardGiven = true;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Consoles/SyndicateOpConsole.cs]: add the brackets and put the assignation in a new line.

File: UnityProject/Assets/Scripts/Systems/Antagonists/AntagManager.cs [hunk 21]
```diff
@@ -25,9 +25,9 @@ public class AntagManager : MonoBehaviour
 		/// <summary>
 		/// All active antagonists
 		/// </summary>
-		private List<SpawnedAntag> ActiveAntags = new List<SpawnedAntag>();
+		private List<SpawnedAntag> activeAntags = new List<SpawnedAntag>();
 
-		public List<SpawnedAntag> CurrentAntags => ActiveAntags;
+		public List<SpawnedAntag> CurrentAntags => activeAntags;
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Systems/Antagonists/AntagManager.cs]: call it ``ActiveAntags``, sorry I thought I was clear enough in my previous comment. It is better to keep the name and only differentiate them by the case style so you know instantly what it is and what access level it has.
